### DISCUSSION TITLE
Tokens in a Multiuser WebApp environment

### DISCUSSION THREAD

**Discussion by pkstacey** [O365/python-o365#821] [General]
Hi all, firstly, I love this library, so complete and efficient. secondly, although I have already successfully implemented it and tested successfully on my development laptop, I'm unsure if when ported to the cloud for multiple concurrent users, whether: 

(a) I  need to change the way the token file is stored? 
(b) Does the o365_token.txt keep getting overwritten by different app users? 

I've looked at the FileSystemTokenBackend feature but not sure if this is necessary. 

Thanks so much!

Patrick

**Comment by pkstacey** [comment]
Hi, I have just implemented the FileSystemTokenBackend as follows in @flask_app.route('/authenticate'). 

I'm not sure if it enables concurrent users to authenticate. Can someone please affirm? My thanks!

`
def authenticate():

    token_backend = FileSystemTokenBackend(token_path=config.__TOKEN_PATH__, token_filename=session['sis_username']+'_token.txt')

    account = Account(credentials, token_backend=token_backend)

    # request a token
    token = account.con.request_token(request.url.replace("http://", "https://", 1), 
                                       state=session['state'],
                                       redirect_uri=callback)

    # if token is True, then authentication was succesful 
    if token:
        return redirect("/feature_menu") #to feature menu
    else:
        return render_template('failure_o365.html')

`

**Comment by alejcas** [comment]
Hi, you should implement a different TokenBackend maybe one that uses a database behind.
Servers can rely on having a file system that is the same for each instance. So unless you are sure the filesystem is the same you should not use FileSystemTokenBackend.

Also check [should_refresh_token](https://github.com/O365/python-o365/blob/master/O365/utils/token.py#L110 ) method from the BaseTokenBackend. This method is intented for this use case.
Read the docstring so you can understand.

**Reply by pkstacey** [reply]
Thanks Janscas! I will look into these valuable suggestions.
Best,

**Reply by pkstacey** [reply]
Hi Janscas. I have now read thoroughly. I am unsure if we are talking about the same thing, ie i may not have been clear. my use case is a multi-user one rather than concurrent connections per user(s). ie my aim is to support multiple users who may only have one connection instance. the app is not intended for constant polling of MS servers but more for periodic ingestion into a dashboard. 

Are we talking about the same use case? If not, possibly all i need to do is implement the database form of token storage?

Thank you very much for your expert advice Janscas.

Patrick
